File: frontend/src/types/live-filter.ts
// Live filter types for BetMines Live Scanner

// Live rule types
export interface LiveStatsRule {
  category: 'live_stats'
  metric: 'goals' | 'total_goals' | 'corners' | 'total_corners'
    | 'shots_on_target' | 'total_shots_on_target' | 'dangerous_attacks'
    | 'total_dangerous_attacks' | 'possession' | 'yellow_cards'
    | 'total_yellow_cards' | 'red_cards' | 'total_red_cards'
  target: 'HOME' | 'AWAY' | 'EITHER' | 'MATCH' | 'FAVORITE' | 'FAVORITE_HOME'
    | 'FAVORITE_AWAY' | 'UNDERDOG' | 'UNDERDOG_HOME' | 'UNDERDOG_AWAY'
    | 'WINNING' | 'LOSING'
  comparator: '=' | '!=' | '>' | '<' | '>=' | '<='
  value?: number // For numeric mode
  compare_to?: string // For advanced mode (e.g., 'AWAY')
}

export interface TeamStateRule {
  category: 'team_state'
  team_state: 'WINNING' | 'LOSING' | 'DRAWING' | 'NOT_WINNING' | 'NOT_LOSING'
  target: 'HOME' | 'AWAY' | 'EITHER'
}

export interface OddsRule {
  category: 'odds'
  market: '1X2' | 'DOUBLE_CHANCE' | 'OVER_UNDER' | 'BTTS' | 'CORNERS'
    | 'HALF_TIME_1X2' | 'HALF_TIME_OVER_UNDER' | 'HALF_TIME_BTTS'
  selection: 'HOME' | 'DRAW' | 'AWAY' | 'OVER' | 'UNDER' | 'YES' | 'NO' | '1X' | '12' | 'X2'
  line?: number // For OU/Corners markets (e.g., 2.5)
  comparator: '=' | '!=' | '>' | '<' | '>=' | '<='
  value: number
}

export interface TimingRule {
  category: 'timing'
  before_minute?: number
  after_minute?: number
  at_minute?: number
}

export interface PreMatchStatsRule {
  category: 'pre_match_stats'
  metric: 'avg_goals_scored' | 'avg_goals_conceded' | 'clean_sheet_pct'
    | 'win_pct' | 'draw_pct' | 'loss_pct' | 'points_per_game'
    | 'ai_home_win_prob' | 'ai_away_win_prob' | 'ai_draw_prob'
    | 'ai_over_2_5_prob' | 'ai_btts_prob' | 'historical_over_2_5_pct'
    | 'historical_btts_pct' | 'historical_1x2_home_pct'
  target: 'HOME' | 'AWAY' | 'EITHER' | 'MATCH'
  comparator: '=' | '!=' | '>' | '<' | '>=' | '<='
  value: number
}

// Union type for all live rule types
export type LiveRule = LiveStatsRule | TeamStateRule | OddsRule | TimingRule | PreMatchStatsRule

// Live filter types
export interface LiveFilter {
  id: number
  user_id: number
  name: string
  description: string | null
  rules: LiveRule[]
  is_active: boolean
  filter_type: 'live' | 'backtest'
  created_at: string
  updated_at: string
}

export interface CreateLiveFilterRequest {
  name: string
  description?: string
  rules: LiveRule[]
  is_active?: boolean
  filter_type: 'live' | 'backtest'
}

export interface UpdateLiveFilterRequest {
  name?: string
  description?: string
  rules?: LiveRule[]
  is_active?: boolean
}

// Live match types
export interface LiveStats {
  corners: {
    home: number
    away: number
    total: number
  }
  shots_on_target: {
    home: number
    away: number
    total: number
  }
  dangerious_attacks: {
    home: number
    away: number
    total: number
  }
  possession: {
    home: number
    away: number
    total: number
  }
  yellow_cards: {
    home: number
    away: number
    total: number
  }
  red_cards: {
    home: number
    away: number
    total: number
  }
  [key: string]: {
    home: number
    away: number
    total: number
  } // Index signature for dynamic access
}

export interface TeamState {
  home: 'WINNING' | 'LOSING' | 'DRAWING' | 'NOT_WINNING' | 'NOT_LOSING'
  away: 'WINNING' | 'LOSING' | 'DRAWING' | 'NOT_WINNING' | 'NOT_LOSING'
  momentum: 'HOME' | 'AWAY' | 'NEUTRAL'
}

export interface AIPredictions {
  home_win_prob: number
  away_win_prob: number
  draw_prob: number
  over_2_5_prob: number
  over_3_5_prob?: number
  btts_prob: number
}

export interface HistoricalStats {
  over_2_5_pct: number
  over_3_5_pct?: number
  btts_yes_pct: number
  home_win_pct: number
}

export interface LiveMatch {
  id: string
  fixture_id: number
  home_team: string
  away_team: string
  home_score: number
  away_score: number
  minute: number
  status: 'LIVE' | 'HALFTIME' | 'FULLTIME' | 'POSTPONED' | 'CANCELLED'
  league: string
  live_stats: LiveStats
  team_state: TeamState
  odds: Record<string, any>
  ai_predictions: AIPredictions
  historical_stats: HistoricalStats
  last_update: string
}

export interface LiveOdds {
  id: number
  fixture_id: number
  market_type: string
  selection: string
  line?: number
  odds: number
  fetched_at: string
}

// Live filter result types (dual-stat comparison)
export interface LiveFilterResult {
  id: number
  filter_id: number
  fixture_id: number
  triggered_at: string
  triggered_minute: number
  notification_value: Record<string, any> // Stats at alert moment
  final_value?: Record<string, any> // Final/current stats
  bet_result: 'PENDING' | 'WIN' | 'LOSS' | 'PUSH'
  resolved_at?: string
  odds_at_trigger?: Record<string, any>
}

// Live scanner stats
export interface LiveScannerStats {
  active_filters: number
  live_matches: number
  alerts_today: number
  success_rate_24h?: number
  avg_odds_today?: number
}

// Live filter backtest types
export interface LiveFilterBacktestRequest {
  filter_id: number
  date_range?: [string, string]
  include_resolved?: boolean
}

export interface LiveFilterBacktestResponse {
  filter_id: number
  total_alerts: number
  resolved_alerts: number
  wins: number
  losses: number
  pushes: number
  success_rate: number
  avg_odds: number
  roi: number
  results: LiveFilterResult[]
}

// Live scanner API responses
export interface LiveScanResponse {
  scan_id: string
  matches_scanned: number
  filters_matched: number
  alerts_generated: number
  scan_duration_ms: number
  timestamp: string
}

export interface LiveFilterTestResponse {
  fixture_id: number
  filter_matches: boolean
  match_data: {
    home_team: string
    away_team: string
    score: string
    minute: number
    status: string
  }
  rule_evaluations: any[]
  timestamp: string
}

// Match history for dual-stat comparison
export interface MatchHistorySnapshot {
  minute: number
  score: string
  corners: string
  possession: string
}

export interface MatchHistoryResponse {
  fixture_id: number
  match: {
    home_team: string
    away_team: string
    current_score: string
    current_minute: number
  }
  history: MatchHistorySnapshot[]
}

// Filter evaluation helpers
export interface FilterEvaluationResult {
  rule: LiveRule
  matches: boolean
  actual_value?: number
  expected_value?: number
  error?: string
}

// KPI cards for live filter results
export interface LiveFilterKPI {
  alerts: number
  success_rate: number
  avg_odds: number
  roi: number
  period: string // '24h', '7d', '30d', 'all'
}

// Real-time updates
export interface LiveMatchUpdate {
  fixture_id: number
  updates: Partial<LiveMatch>
  timestamp: string
}

export interface LiveFilterAlert {
  filter_id: number
  fixture_id: number
  alert_type: 'NEW_MATCH' | 'STAT_UPDATE' | 'ODDS_CHANGE' | 'MATCH_END'
  message: string
  timestamp: string
}

File: backend/app/schemas/live_filter.py
"""Live filter schemas for request/response validation."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


# Base rule schemas for different categories
class LiveStatsRule(BaseModel):
    """Live Stats rule (Category A) - In-game events and stats."""

    category: Literal["live_stats"] = Field(..., description="Rule category")
    metric: Literal[
        "goals", "total_goals", "corners", "total_corners",
        "shots_on_target", "total_shots_on_target", "dangerous_attacks",
        "total_dangerous_attacks", "possession", "yellow_cards",
        "total_yellow_cards", "red_cards", "total_red_cards"
    ] = Field(..., description="Live stat metric")
    target: Literal[
        "HOME", "AWAY", "EITHER", "MATCH", "FAVORITE", "FAVORITE_HOME",
        "FAVORITE_AWAY", "UNDERDOG", "UNDERDOG_HOME", "UNDERDOG_AWAY",
        "WINNING", "LOSING"
    ] = Field(..., description="Target for the stat")
    comparator: Literal["=", "!=", ">", "<", ">=", "<="] = Field(..., description="Comparison operator")
    value: float | None = Field(None, description="Value for numeric mode")
    compare_to: str | None = Field(None, description="Target for advanced mode (e.g., 'AWAY')")


class TeamStateRule(BaseModel):
    """Team State rule (Category A extension) - Win/Draw/Loss status."""

    category: Literal["team_state"] = Field(..., description="Rule category")
    team_state: Literal["WINNING", "LOSING", "DRAWING", "NOT_WINNING", "NOT_LOSING"] = Field(
        ..., description="Team state condition"
    )
    target: Literal["HOME", "AWAY", "EITHER"] = Field(..., description="Which team to check")


class OddsRule(BaseModel):
    """Odds rule (Category C) - Live and pre-match odds."""

    category: Literal["odds"] = Field(..., description="Rule category")
    market: Literal[
        "1X2", "DOUBLE_CHANCE", "OVER_UNDER", "BTTS", "CORNERS",
        "HALF_TIME_1X2", "HALF_TIME_OVER_UNDER", "HALF_TIME_BTTS"
    ] = Field(..., description="Betting market")
    selection: Literal["HOME", "DRAW", "AWAY", "OVER", "UNDER", "YES", "NO", "1X", "12", "X2"] = Field(
        ..., description="Market selection"
    )
    line: float | None = Field(None, description="Line for OU/Corners markets (e.g., 2.5)")
    comparator: Literal["=", "!=", ">", "<", ">=", "<="] = Field(..., description="Comparison operator")
    value: float = Field(..., description="Odds value to compare")


class TimingRule(BaseModel):
    """Timing rule (Category D) - Match minute filters."""

    category: Literal["timing"] = Field(..., description="Rule category")
    before_minute: int | None = Field(None, ge=1, le=120, description="Trigger before this minute")
    after_minute: int | None = Field(None, ge=1, le=120, description="Trigger after this minute")
    at_minute: int | None = Field(None, ge=1, le=120, description="Trigger at this exact minute")


class PreMatchStatsRule(BaseModel):
    """Pre-Match Stats rule (Category E) - Historical and AI probabilities."""

    category: Literal["pre_match_stats"] = Field(..., description="Rule category")
    metric: Literal[
        "avg_goals_scored", "avg_goals_conceded", "clean_sheet_pct",
        "win_pct", "draw_pct", "loss_pct", "points_per_game",
        "ai_home_win_prob", "ai_away_win_prob", "ai_draw_prob",
        "ai_over_2_5_prob", "ai_btts_prob", "historical_over_2_5_pct",
        "historical_btts_pct", "historical_1x2_home_pct"
    ] = Field(..., description="Pre-match stat metric")
    target: Literal["HOME", "AWAY", "EITHER", "MATCH"] = Field(..., description="Target for the stat")
    comparator: Literal["=", "!=", ">", "<", ">=", "<="] = Field(..., description="Comparison operator")
    value: float = Field(..., description="Value to compare")


# Union type for all live rule types
LiveRule = (
    LiveStatsRule | TeamStateRule | OddsRule | TimingRule | PreMatchStatsRule
)


class LiveFilterCreate(BaseModel):
    """Schema for creating a new live filter."""

    name: str = Field(..., min_length=1, max_length=100, description="Filter name")
    description: str | None = Field(None, max_length=500, description="Filter description")
    rules: list[LiveRule] = Field(
        ..., min_length=1, max_length=10, description="Live filter conditions (max 10)"
    )
    is_active: bool = Field(True, description="Whether filter is active")
    filter_type: Literal["live", "backtest"] = Field(..., description="Filter type")


class LiveFilterUpdate(BaseModel):
    """Schema for updating an existing live filter."""

    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    rules: list[LiveRule] | None = Field(None, min_length=1, max_length=10)
    is_active: bool | None = None


class LiveMatchResponse(BaseModel):
    """Schema for live match response."""

    id: int
    fixture_id: int
    status: str
    minute: int
    home_score: int
    away_score: int
    live_stats: dict[str, Any]
    home_team_state: str | None
    away_team_state: str | None
    momentum: str | None
    ai_predictions: dict[str, Any] | None
    historical_stats: dict[str, Any] | None
    last_update: datetime

    model_config = {"from_attributes": True}


class LiveOddsResponse(BaseModel):
    """Schema for live odds response."""

    id: int
    fixture_id: int
    market_type: str
    selection: str
    line: float | None
    odds: float
    fetched_at: datetime

    model_config = {"from_attributes": True}


class LiveFilterResultResponse(BaseModel):
    """Schema for live filter result response with dual-stat comparison."""

    id: int
    filter_id: int
    fixture_id: int
    triggered_at: datetime
    triggered_minute: int
    notification_value: dict[str, Any]  # Stats at alert moment
    final_value: dict[str, Any] | None  # Final/current stats
    bet_result: str
    resolved_at: datetime | None
    odds_at_trigger: dict[str, Any] | None

    model_config = {"from_attributes": True}


class LiveScannerStatsResponse(BaseModel):
    """Schema for live scanner statistics."""

    active_filters: int
    live_matches: int
    alerts_today: int
    success_rate_24h: float | None
    avg_odds_today: float | None


class LiveFilterBacktestRequest(BaseModel):
    """Schema for live filter backtest request."""

    filter_id: int
    date_range: tuple[str, str] | None = Field(None, description="Date range for backtest")
    include_resolved: bool = Field(True, description="Include resolved results")


class LiveFilterBacktestResponse(BaseModel):
    """Schema for live filter backtest response."""

    filter_id: int
    total_alerts: int
    resolved_alerts: int
    wins: int
    losses: int
    pushes: int
    success_rate: float
    avg_odds: float
    roi: float
    results: list[LiveFilterResultResponse]
